### ISSUE TITLE
Webpack excluding dependencies starting with "electron-"

### ISSUE THREAD

**Issue by robido** [oliverfindl/electron-vue-boilerplate#6]
**Note:** the issue is the same with package electron-fetch or electron-log, so I don't think its an issue related to the package.

I made a minimal change to show the problem:
https://github.com/robido/electron-vue-boilerplate/commit/c7c76bdc0de851d94217b23f925eb7a21c8f88df

It all works as expected when the app is run in development mode:
```
Terminal 1: npm run dev
Terminal 2: npm run start
```

However, when packing `npm run pack`, and running App.exe it shows this error:

![image](https://user-images.githubusercontent.com/7013911/115819150-461e9f80-a3cc-11eb-8913-37f342148c1c.png)

**Comment by robido** [comment]
For reference, I am using node v14.16.1 (npm 6.14.12)

**Comment by oliverfindl** [comment]
Hello,

this behavior is due to building / bundling configuration. You could fiddle with `build.files` option in [package.json](https://github.com/oliverfindl/electron-vue-boilerplate/blob/02bf7b192a27643e3b60087d0756f638544aa412/package.json):

https://github.com/oliverfindl/electron-vue-boilerplate/blob/02bf7b192a27643e3b60087d0756f638544aa412/package.json#L30-L33

Change to:

```javascript
"files": [
	"dist/**/*"
],
```

[ [docs](https://www.electron.build/configuration/contents.html#files) ]

Or just pass desired packages into [webpack-node-externals](https://github.com/liady/webpack-node-externals) configuration as `allowList` via `externals` option in [webpack-electron-app.config.js](https://github.com/oliverfindl/electron-vue-boilerplate/blob/02bf7b192a27643e3b60087d0756f638544aa412/webpack-electron-app.config.js):

https://github.com/oliverfindl/electron-vue-boilerplate/blob/02bf7b192a27643e3b60087d0756f638544aa412/webpack-electron-app.config.js#L44

Change to:

```javascript
externals: [
	nodeExternals({
		allowlist: [ "electron-log" ]
	})
],
```

[ [docs](https://github.com/liady/webpack-node-externals#configuration) ]

Tested both solutions and both worked for me. At this point, its up to you to explore built `asar` package to see, which packages from `node_modules` are included in your app and therefore, which solution suits your use-case better.

If you have more questions, feel free to ask here.

Thanks.

**Comment by robido** [comment]
Thanks for looking into it! You put me in the right path.

Your proposed solution worked with electron-log and electron-fetch, but in my project, here are my dependencies:

```
"dependencies": {
  "@grpc/grpc-js": "^1.2.12",
  "@panter/vue-i18next": "^0.15.2",
  "bootstrap": "^4.6.0",
  "bootstrap-vue": "^2.21.2",
  "d3": "^6.7.0",
  "d3fc": "^15.1.2",
  "electron-fetch": "^1.7.3",
  "electron-log": "^4.3.4",
  "google-protobuf": "^3.15.8",
  "humanize-duration": "^3.25.1",
  "i18next": "^20.2.1",
  "js-time-diff": "^1.0.0",
  "lodash": "^4.17.21",
  "mathjs": "^9.3.2",
  "popper.js": "^1.16.1",
  "timestamp-nano": "^1.0.0",
  "vue": "^2.6.12",
  "vue-json-print": "^0.3.2",
  "vue-router": "^3.5.1",
  "vue-split-panel": "^1.0.4",
  "vuex": "^3.6.2",
  "zip-local": "^0.3.4"
 },
```

1) Solution 1 (remove line `"!node_modules${/*}"`): works, but built asar jumped from 3 MB to 80 MB. When not using electron-log and electron-fetch, the app starts normally, all the other dependencies are included in the 3 MB, so that 80 MB seems unnecessary.

2) Solution 2 (`allowlist: [ "electron-log" ]`): worked for electron-log and electron-fetch, but I got the error of cannot find module of other sub-modules (electron-fetch depends on encoding, and encoding depends on iconv-lite, and inconv-lite depends on safer-buffer), which meant all submodules needed to be allowed: `allowlist: [ "electron-log", "electron-fetch", "encoding", "iconv-lite", "safer-buffer" ]`. Even after doing this, it did not work for me, I started getting errors on packages that worked before (vue not found).

3) Solution 3: I tried to completely remove line `externals: [ nodeExternals() ]` (and remove it from devDependencies) which seems to work. The asar is 3MB, and all modules load correctly. To me this solution closes the issue.

Question: is nodeExternals needed or it's safe to remove? According to their documentation it is to avoid bundling node_modules for backends, but a electron-vue app is not running in a server.

**Comment by oliverfindl** [comment]
Hello,

`externals` are not strictly required to be defined, but it helps your app to keep smaller size (although, not most crucial point of Electron app, which its huge by itself). Issue is not with vue-app, but with electron-app, which is server for vue-app and which needs its dependencies (that are not bundled), therefore it needs load dependencies from `node_modules` directory of your app. I'm not 100% sure, but I think with removal of `externals` option, all your dependencies get copied at build time of electron-app and at same time, most of them also gets bundled into vue-app, so you probably have most of them 2 times included in your app. `externals` option should serve as filter to which dependencies you want ignore (by default, with [webpack-node-externals](https://github.com/liady/webpack-node-externals), all of them) in your project (electron-app in this case).

I hope this helps, although I'm not 100% sure if this is correct, because I didn't make any Electron app for long time. I'm only updating dependencies of this package, so maybe something might have changed from my last experience with Electron.

Thanks.

EDIT:

I just tested with commented `externals` and it doesn't include dependencies 2 times (this was some other issue I had back then). I remember, I had to add `externals` because some issue, but can't remember what it was exactly (something with internal node http module or socket.io package).

Thanks.
